File: library/src/main/java/com/getstream/sdk/chat/model/Event.java
package com.getstream.sdk.chat.model;

import androidx.room.Ignore;

import com.getstream.sdk.chat.enums.EventType;
import com.getstream.sdk.chat.interfaces.UserEntity;
import com.getstream.sdk.chat.rest.Message;
import com.getstream.sdk.chat.rest.User;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import java.util.Date;

/**
 * An event
 */
public class Event implements UserEntity {
    @SerializedName("connection_id")
    @Expose
    private String connectionId;

    @SerializedName("cid")
    @Expose
    private String cid;

    @SerializedName("client_id")
    @Expose
    private String clientId;

    @SerializedName("type")
    @Expose
    private String type;

    @SerializedName("user")
    @Expose
    private User user;

    @SerializedName("me")
    @Expose
    private User me;

    @SerializedName("member")
    @Expose
    private Member member;

    @SerializedName("message")
    @Expose
    private Message message;

    @SerializedName("reaction")
    @Expose
    private Reaction reaction;

    @SerializedName("channel")
    @Expose
    private Channel channel;

    @SerializedName("total_unread_count")
    @Expose
    private Number totalUnreadCount;

    @SerializedName("unread_channels")
    @Expose
    private Number unreadChannels;

    @SerializedName("watcher_count")
    @Expose
    private Number watcherCount;

    @SerializedName("created_at")
    @Expose
    private Date createdAt;

    @SerializedName("clear_history")
    @Expose
    private Boolean clearHistory;

    @Ignore
    private Date receivedAt;

    private boolean online;

    public Event() {
    }

    public Event(String type) {
        this.type = type;
    }

    public Event(boolean online) {
        this.online = online;
        setType(EventType.CONNECTION_CHANGED);
    }

    public String getConnectionId() {
        return connectionId;
    }

    public void setConnectionId(String connectionId) {
        this.connectionId = connectionId;
    }

    public boolean isChannelEvent() {
        return cid != null && !cid.equals("*");
    }

    public String getCid() {
        return cid;
    }

    public EventType getType() {
        return EventType.findByString(type);
    }

    public void setType(String type) {
        this.type = type;
    }

    public void setType(EventType type) {
        this.type = type.label;
    }

    public User getUser() {
        return user;
    }

    public User getMe() {
        return me;
    }

    public boolean isAnonymous() {
        if (me != null) {
            return me.getId().equals("!anon");
        }
        return true;
    }

    public Member getMember() {
        return member;
    }

    public Message getMessage() {
        return message;
    }

    public void setMessage(Message message) {
        this.message = message;
    }

    public Reaction getReaction() {
        return reaction;
    }

    public Channel getChannel() {
        return channel;
    }

    public Number getTotalUnreadCount() {
        return totalUnreadCount;
    }

    public Number getUnreadChannels() {
        return unreadChannels;
    }

    public Number getWatcherCount() {
        return watcherCount;
    }

    public Date getCreatedAt() {
        return createdAt;
    }

    public Boolean getOnline() {
        return online;
    }

    public String getClientId() {
        return clientId;
    }

    public void setClientId(String clientId) {
        this.clientId = clientId;
    }

    public String getUserId() {
        if (user == null) {
            return null;
        }
        return user.getId();
    }

    public Date getReceivedAt() {
        return receivedAt;
    }

    public void setReceivedAt(Date receivedAt) {
        this.receivedAt = receivedAt;
    }

    public Boolean getClearHistory() {
        return clearHistory;
    }

    public void setClearHistory(Boolean clearHistory) {
        this.clearHistory = clearHistory;
    }
}

File: app/src/main/java/io/getstream/chat/android/core/poc/library/Event.kt
package io.getstream.chat.android.core.poc.library

import com.google.gson.annotations.SerializedName
import io.getstream.chat.android.core.poc.library.utils.UndefinedDate
import java.util.*


data class Event(
    private var type: String = ""
) : UserEntity {

    val cid: String = ""
    val user: User? = null
    val me: User? = null
    val member: Member? = null
    var message: Message? = null
    val reaction: Reaction? = null
    val channel: Channel? = null
    var receivedAt: Date = UndefinedDate
    var online = false

    @SerializedName("connection_id")
    var connectionId: String = ""
    @SerializedName("client_id")
    var clientId: String = ""
    @SerializedName("total_unread_count")
    val totalUnreadCount: Number = 0
    @SerializedName("unread_channels")
    val unreadChannels: Number = 0
    @SerializedName("watcher_count")
    val watcherCount: Number = 0
    @SerializedName("created_at")
    val createdAt: Date = UndefinedDate
    @SerializedName("clear_history")
    var clearHistory: Boolean = false

    constructor(type: EventType) : this(type.label) {
        this.type = type.label
    }

    constructor(type: EventType, online: Boolean) : this(type.label) {
        this.online = online
        this.type = type.label
    }

    val isChannelEvent: Boolean
        get() = cid != "*"

    fun getType(): EventType {
        return EventType.values().firstOrNull {
            it.label == type
        } ?: EventType.UNKNOWN
    }

    val isAnonymous: Boolean
        get() = if (me != null) {
            me.id == "!anon"
        } else true

    override fun getUserId(): String {
        return user!!.id
    }



}